src/main/java/build/buf/protovalidate/Config.java

```diff
@@ -80,13 +128,67 @@ public Builder setDisableLazy(boolean disableLazy) {
       return this;
     }
 
+    /**
+     * Set the type registry for reparsing protobuf messages. This option should be set alongside
+     * setExtensionRegistry to allow dynamic resolution of predefined rule extensions. It should be
+     * set to a TypeRegistry with all the message types from your file descriptor set registered. By
+     * default, if any unknown field constraints are found, compilation of the constraints will
+     * fail; use setAllowUnknownFields to control this behavior.
+     *
+     * <p>Note that the message types for any extensions in setExtensionRegistry must be present in
+     * the typeRegistry, and have an exactly-equal Descriptor. If the type registry is not set, the
+     * extension types in the extension registry must have exactly-equal Descriptor types to the
+     * protovalidate built-in messages. If these conditions are not met, extensions will not be
+     * resolved as expected. These conditions will be met when constructing a TypeRegistry and
+     * ExtensionRegistry using information from the same file descriptor sets.
+     *
+     * @param typeRegistry the type registry to use
+     * @return this builder
+     */
+    public Builder setTypeRegistry(TypeRegistry typeRegistry) {
```

Comment:
It might actually be possible to do this but it's sort of complicated since we're jumping back and forth between user protos and our own. But actually, that's not my main concern. I have really two problems:

1. `TypeRegistry` is immutable. Once it is constructed, it is done. Therefore, incrementally building a `TypeRegistry` is always inefficient. We can, of course, build our own `TypeRegistry` which is essentially just a `Map`, but that smells. And it gets worse, because,
2. The behavior of conflicting file descriptors will _depend on the order in which you try to validate messages_. To me this is just too confusing.

Bear in mind that this situation (unknown extensions) should realistically only be a problem for users that are dealing with dynamic file descriptor sets. In many cases users that are dealing with dynamic descriptor sets might have to deal with multiple different dynamic descriptor sets, especially if they're doing some kind of multi-tenant service. By exposing the `TypeRegistry` we make it explicit: The user controls what resolves to what `Descriptor`, and you need a new `Validator` for a new `TypeRegistry`.

I can definitely see a case for us providing better helper functions. However, the helper functions for this would really not be terribly protovalidate-specific and so I opted to keep the helpers in our private code in the conformance test suite, which users could conceivably copy if they wanted our behavior. Users who are dealing with dynamic protocol buffers likely need to build an `ExtensionRegistry` and/or a `TypeRegistry` anyhow, and probably pass it to other libraries or use it in their own code.

Of course I am open to trying to improve this situation but I opted to do things the dumb way, because it's already pretty confusing to begin with, and adding more magic on top of it seems like it is just not the greatest user experience.

Reply:
It's a good point that we could just provide our own type since we're not actually providing the `TypeRegistry` to any API that requires it.

> The behavior of conflicting file descriptors will depend on the order in which you try to validate messages. To me this is just too confusing.

If we do it right, there is no such thing as conflicting descriptors. A single descriptor is not valid if it has conflicts within its own transitive dependency graph. So we really want a tree of maps -- each node has a map of the messages in that file, and then it has tree links to its imports. So finding a type could require O(n) time (where n is number of files in transitive graph) instead of O(1), but it wouldn't be subject to conflicts.

Sadly, we can't really do this with extension types since `ExtensionRegistry` is a concrete type that _is_ required by APIs we're using.

Note that I wasn't suggesting incrementally building a single shared registry but rather to build a new registry _per file descriptor_. I understand that could be a perf issue, but since it's only done once per file per cache (and there will likely be a single cache for most programs plus fixed number of files with generated message types) it is likely to be perfectly acceptable.

Reply:
Unfortunately, I just realized that our own tests illuminate a case where this won't work.

Consider the following:

```java
    ExtensionRegistry registry = ExtensionRegistry.newInstance();
    registry.add(isIdent);
    TypeRegistry typeRegistry =
        TypeRegistry.newBuilder().add(isIdent.getDescriptor().getContainingType()).build();
    Config config = Config.newBuilder().setTypeRegistry(typeRegistry).setExtensionRegistry(registry).build();
```

This obviously works just fine because the `TypeRegistry` and the `ExtensionRegistry` are synced with each-other. However... if we get rid of the `TypeRegistry`, this test can never work!

This is because this test does not actually go across two different `Descriptor` types like one might assume. It goes across _three_, each with their own distinct `hashCodes`.

- The first one is `build.buf.validate.StringRules`.
- The second one is `com.example.imports.buf.validate.StringRules`. We have a second copy of the gencode to specifically test this case (dynamic proto of separate gencode.)
- The third one is a dynamically-resolved `com.validate.imports.buf.validate.StringRules` - crucially, extensions aren't resolved in nested types.

Our code already intends to handle this case (dynamic message parsed from potentially distinct file descriptor; we test it for other cases that don't involve extensions). This puts us in a tricky position, because we really want to reparse the extensions using the extension types, _not_ the types that happen to be attached to the descriptor of a message, which may be dynamic or have a different gencode than whoever registered the extensions.

The only other alternative I can think of to explicitly having the user register a `TypeRegistry` is if we enumerate their `ExtensionRegistry` to find the types instead, but it seems like it just keeps getting worse the harder you try:
- There's no great way to enumerate all extensions in an `ExtensionRegistry` so what you actually have to do is enumerate the extensions of each possible _message_ that you might want to query extensions for.
- There's no guarantee that there's only one `Descriptor` for a given full name stored in the `ExtensionRegistry` - it will happy return a list with extensions that have many distinct `Descriptor` types. We would probably need to throw an exception in this case since there's not much reasonable to do with it.

At this point I'm torn. The test case is indeed a bit convoluted, but it's convoluted on purpose. With the `TypeRegistry` it works for fairly obvious reasons. Without it, it doesn't work for reasons that I found very unintuitive.

What would be nice is if we could just have the `ExtensionRegistry` key on the full name of the descriptor instead of the descriptor identity, but after looking around for too long I've determined there's nothing we can really do on that front other than things that I'm pretty sure we shouldn't do.